File: project/black_scholes.py
"""
Julius Olson
IE 420 Project

Black-Scholes
"""

from scipy.stats import norm
from math import log, sqrt, exp

def N_cdf(x):
	return norm(0, 1).cdf(x)

def black_scholes(Option, K, T, S0, sigma, r, q, N, Exercise):
	d1 = (log(S0 / K) + (r-q + 0.5 * sigma**2) * T) / (sigma * sqrt(T))
	d2 = d1 - sigma * sqrt(T)

	if Option == "C":
		return S0 * exp(-q*T) * N_cdf(d1) - K*exp(-r*T) * N_cdf(d2)
	elif Option == "P":
		return -S0 * exp(-q*T) * N_cdf(-d1) + K*exp(-r*T) * N_cdf(-d2)
	else:
		raise TypeError

File: project/src/CPP/black_scholes.cpp
/*
	IE420 Project
	Julius Olson
	Black & Scholes
*/

#include <cmath>
#include "binomial.h"


double n_cdf(double x) {
	return 0.5 * erfc(-x * sqrt(0.5));
}

double black_scholes(option &opt) {
	double d1, d2;

	d1 = (log(opt.S0 / opt.K) + (opt.r - opt.q + 0.5 * pow(opt.sigma, 2)) * opt.T) / (opt.sigma * sqrt(opt.T));
	d2 = d1 - opt.sigma * sqrt(opt.T);

	if (opt.option == C) {
		return opt.S0 * exp(-opt.q * opt.T) * n_cdf(d1) - opt.K * exp(-opt.r * opt.T) * n_cdf(d2);
	} else {
		return -opt.S0 * exp(-opt.q * opt.T) * n_cdf(d1) + opt.K * exp(-opt.r * opt.T) * n_cdf(d2);
	}
}